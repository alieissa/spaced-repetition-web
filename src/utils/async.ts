/** @format */
/**
 * Credit goes to https://medium.com/@fillopeter/pattern-matching-with-typescript-done-right-94049ddd671c
 * For the pattern matching code
 */
export type Untriggered = {
  type: 'Untriggered'
}
export function Untriggered(): Untriggered {
  return {
    type: 'Untriggered',
  }
}

const t: Optional<Answer, 'easiness' | 'quality' | 'interval'> = {
  id: '',
  content: 'fsdfds',
}
export type Loading<V> = {
  type: 'Loading'
  value: V
}
export function Loading<V>(v: V): Loading<V> {
  return {
    type: 'Loading',
    value: v,
  }
}

export type Failure<E> = {
  type: 'Failure'
  value: E
}
export function Failure<E>(e: E): Failure<E> {
  return {
    type: 'Failure',
    value: e,
  }
}

export type Success<V> = {
  type: 'Success'
  value: V
}
export function Success<V>(v: V): Success<V> {
  return {
    type: 'Success',
    value: v,
  }
}

export type Async<L = unknown, F = unknown, S = unknown> =
  | Untriggered
  | Loading<L>
  | Failure<F>
  | Success<S>

export type AsyncStatus<R, E, A> = Async<R, E, A>['type']

export type AsyncMap<R, E, A, U> = {
  [K in AsyncStatus<R, E, A>]: U extends { type: K } ? U : never
}

export type AsyncStatusMap<R, E, A> = AsyncMap<R, E, A, Async<R, E, A>>
export type AsyncPattern<T, R, E, A> = {
  [K in keyof AsyncStatusMap<R, E, A>]: (a: AsyncStatusMap<R, E, A>[K]) => T
}

export function match<R, E, A>(
  e: Async<R, E, A>,
): <T>(pattern: AsyncPattern<T, R, E, A>) => T {
  return <T>(pattern: AsyncPattern<T, R, E, A>) => pattern[e.type](e as any)
}
