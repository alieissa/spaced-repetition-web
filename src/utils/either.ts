/** @format */

/**
 * Credit goes to https://medium.com/@fillopeter/pattern-matching-with-typescript-done-right-94049ddd671c
 * For the pattern matching code
 */
export type Right<V> = Readonly<{
  type: 'Right'
  value: V
}>
export function Right<V>(v: V): Right<V> {
  return {
    type: 'Right',
    value: v,
  }
}

export type Left<E = Error> = Readonly<{
  type: 'Left'
  value: E
}>
export function Left<E>(e: E): Left<E> {
  return {
    type: 'Left',
    value: e,
  }
}

export type Either<L, R> = Left<L> | Right<R>

export type EitherType<L, R> = Either<L, R>['type']
export type EitherMap<L, R, U> = {
  [K in EitherType<L, R>]: U extends { type: K } ? U : never
}
export type EitherTypeMap<L, R> = EitherMap<L, R, Either<L, R>>
export type EitherPattern<T, L, R> = {
  [K in keyof EitherTypeMap<L, R>]: (either: EitherTypeMap<L, R>[K]) => T
}

export function match<L, R>(
  e: Either<L, R>,
): <T>(pattern: EitherPattern<T, L, R>) => T {
  return <T>(pattern: EitherPattern<T, L, R>) => pattern[e.type](e as any)
}
