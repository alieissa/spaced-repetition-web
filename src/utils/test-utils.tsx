/** @format */

import { combineReducers, configureStore } from '@reduxjs/toolkit'
import { render } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import { RouterProvider, createMemoryRouter } from 'react-router-dom'
import decksReducer from 'src/modules/decks/decks.reducer'

// As a basic setup, import your same slice reducers
import { loginReducer } from 'src/modules/login'
import { signupReducer } from 'src/modules/signup'
import { verificationReducer } from 'src/modules/verification'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
// interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
//   preloadedState?: PreloadedState<RootState>
//   store?: AppStore
// }

const reducers = combineReducers({
  login: loginReducer,
  signup: signupReducer,
  verification: verificationReducer,
  decks: decksReducer,
})

const withRedux = (ui: React.ReactElement) => {
  const store = configureStore({ reducer: reducers })
  return <Provider store={store}>{ui}</Provider>
}
const withRouter = (ui: React.ReactElement) => {
  const router = createMemoryRouter([{ element: ui, path: '/' }], {
    initialEntries: ['/'],
    initialIndex: 1,
  })

  return <RouterProvider router={router} />
}
export function renderWithProviders(ui: React.ReactElement) {
  // Return an object with the store and all of RTL's query functions
  return render(withRedux(withRouter(ui)))
}

export const flushPromises = () =>
  new Promise((resolve) => setTimeout(resolve, 0))
