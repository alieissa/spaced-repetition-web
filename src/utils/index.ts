/** @format */

import * as async from './async'
import * as either from './either'

export { async }
export { either }
