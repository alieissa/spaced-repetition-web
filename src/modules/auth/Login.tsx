/** @format */

import { FormikState, useFormik } from 'formik'
import { ChangeEvent, useEffect } from 'react'
import { Link, useNavigate } from 'react-router-dom'
import { Container, Form, Loader, Message } from 'semantic-ui-react'
import { SPButton, SPCheckbox, SPHeader, SPInput, SPText } from 'src/components'
import { styles } from 'src/styles'
import { async } from 'src/utils'
import * as Yup from 'yup'
import useLogin from './auth.hooks'
import { NAuth } from './auth.types'

const LoginSchema = Yup.object().shape({
  email: Yup.string().email('Invalid email').required('Required'),
  password: Yup.string().required('Required'),
})
type Props = {
  form: FormikState<NAuth.UserForm>
  onLogin: (data: NAuth.UserForm) => void
  onChangeEmail: (e: ChangeEvent<HTMLInputElement>) => void
  onChangePassword: (e: ChangeEvent<HTMLInputElement>) => void
  onChangeRememberMe: VoidFunction
}
const LoginComponent = (props: Props) => {
  const emailError = props.form.touched.email && !!props.form.errors.email
  const passwordError =
    props.form.touched.password && !!props.form.errors.password

  return (
    <>
      <Form
        onSubmit={() => props.onLogin(props.form.values)}
        className="flex-column bordered p-1r"
      >
        <Form.Field error={emailError}>
          <label>Email</label>
          <SPInput
            data-testid="login-form-email-input"
            title="email"
            disabled={props.form.isSubmitting}
            placeholder="Email"
            value={props.form.values.email}
            onChange={props.onChangeEmail}
          />
          {emailError && (
            <SPText
              data-testid="login-form-email-error"
              style={styles['mt-0.5r']}
            >
              {props.form.errors.email}
            </SPText>
          )}
        </Form.Field>
        <Form.Field error={passwordError}>
          <label>Password</label>
          <SPInput
            data-testid="login-form-password-input"
            title="password"
            disabled={props.form.isSubmitting}
            placeholder="Password"
            type="password"
            value={props.form.values.password}
            onChange={props.onChangePassword}
          />
          <div className="flex-row-reverse" style={styles['mt-0.5r']}>
            <Link
              data-testid="login-form-forgot-password-link"
              to="/forgot-password"
            >
              Forgot password?
            </Link>
          </div>
          {passwordError && (
            <SPText
              data-testid="login-form-password-error"
              style={styles['mt-0.5r']}
            >
              {props.form.errors.password}
            </SPText>
          )}
        </Form.Field>
        <footer className="align-center justify-space-between">
          <div className="align-center">
            <SPCheckbox
              checked={props.form.values.rememberMe}
              onChange={props.onChangeRememberMe}
            />
            <span className="ml-fourth-r">Remember Me?</span>
          </div>
          <SPButton
            type="submit"
            disabled={props.form.isSubmitting}
            className="align-self-end"
          >
            Login
          </SPButton>
        </footer>
      </Form>
    </>
  )
}

function Login() {
  const navigate = useNavigate()
  const [status, login, resetLogin] = useLogin()

  const form = useFormik({
    initialValues: {
      password: '',
      email: '',
      rememberMe: true,
    },
    validationSchema: LoginSchema,
    onSubmit: (data: NAuth.UserForm) => login(data),
  })

  // Very important: Makes sure that when login is mounted
  // as a result of a redirect from the app the login status
  // is not success. Without this cleanup code we would have
  // an infinite loop
  useEffect(() => {
    return () => resetLogin()
    //eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  useEffect(() => {
    form.setSubmitting(status.type === 'Loading')
    if (status.type !== 'Success') {
      return
    }
    navigate('/')
    //eslint-disable-next-line react-hooks/exhaustive-deps
  }, [status.type])

  const handleChangeEmail = (e: ChangeEvent<HTMLInputElement>) => {
    form.setFieldValue('email', e.target.value)
  }

  const handleChangePassword = (e: ChangeEvent<HTMLInputElement>) => {
    form.setFieldValue('password', e.target.value)
  }

  const handleChangeRememberMe = () => {
    form.setFieldValue('rememberMe', !form.values.rememberMe)
  }

  const handleSignupClick = () => navigate('/signup')

  const commonProps = {
    form,
    onChangeEmail: handleChangeEmail,
    onChangePassword: handleChangePassword,
    onChangeRememberMe: handleChangeRememberMe,
    onLogin: form.submitForm,
  }

  return (
    <Container>
      <div className="justify-space-between bordered p-1r mb-1r">
        <SPHeader as="h1">Spaced Reps</SPHeader>
        <div>
          <SPButton
            data-testid="login-form-signup-button"
            size="medium"
            color="blue"
            onClick={handleSignupClick}
          >
            Signup
          </SPButton>
        </div>
      </div>
      {async.match(status)({
        Untriggered: () => (
          <div data-testid="login-form">
            <LoginComponent {...commonProps} />
          </div>
        ),
        Loading: () => (
          <div data-testid="login-form-loading">
            <Loader active />
            <LoginComponent {...commonProps} />
          </div>
        ),
        Failure: () => (
          <div data-testid="login-form-error">
            <Message negative>
              <Message.Header>Login failed</Message.Header>
              <p>Please make sure email and/or password are correct</p>
            </Message>
            <LoginComponent {...commonProps} />
          </div>
        ),
        Success: () => null,
      })}
    </Container>
  )
}

export default Login
