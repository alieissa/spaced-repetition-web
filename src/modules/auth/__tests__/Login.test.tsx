/** @format */

import '@testing-library/jest-dom'
import {
  act,
  screen,
  waitFor,
  waitForElementToBeRemoved,
} from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { rest } from 'msw'
import { setupServer } from 'msw/node'
import { renderWithProviders } from 'src/utils/test-utils'
import Login from '../Login'
const mockNavigate = jest.fn(() => ({}))
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}))

const loginUrl = `${process.env.REACT_APP_API_ENDPOINT}/users/login`

export const handlers = [
  rest.post(loginUrl, (__, res, ctx) => {
    return res(ctx.json({ token: 'dummytoken1234' }))
  }),
]
// const formFields = ['email', 'password'] as const

const setupForm = async (fields: ('email' | 'password')[]) => {
  const user = userEvent.setup()

  if (fields.includes('email')) {
    const emailInput = screen.getByTestId('login-form-email-input')
    await act(() => user.type(emailInput, 'johnnyb1@gmail.com'))
  }

  if (fields.includes('password')) {
    const passwordInput = screen.getByTestId('login-form-password-input')
    await act(() => user.type(passwordInput, 'heythere!'))
  }
}

const server = setupServer(...handlers)
beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

describe('Login', () => {
  describe('view', () => {
    it('should render correctly', () => {
      const { asFragment } = renderWithProviders(<Login />)
      expect(asFragment()).toMatchSnapshot()
    })
    test('display initial', async () => {
      renderWithProviders(<Login />)
      expect(screen.getByTestId('login-form')).toBeInTheDocument()
    })

    test('display loading', async () => {
      // Assemble
      renderWithProviders(<Login />)
      await setupForm(['email', 'password'])

      // Act
      const loginBtn = screen.getByRole('button', { name: 'Login' })
      await act(() => loginBtn.click())

      // Assert
      const testId = 'login-form-loading'
      expect(await screen.findByTestId(testId)).toBeInTheDocument()
      // See https://davidwcai.medium.com/react-testing-library-and-the-not-wrapped-in-act-errors-491a5629193b
      await waitForElementToBeRemoved(() => screen.queryByTestId(testId))
    })

    test('display failure', async () => {
      // Assemble
      server.use(
        rest.post(loginUrl, (__, res, ctx) =>
          res(ctx.json({ email: 'invalidemail' }), ctx.status(422)),
        ),
      )
      renderWithProviders(<Login />)
      await setupForm(['email', 'password'])

      // Act
      const loginBtn = screen.getByRole('button', { name: 'Login' })
      await act(() => loginBtn.click())

      // Assert
      const testId = 'login-form-error'
      expect(await screen.findByTestId(testId)).toBeInTheDocument()
    })

    test('display success', async () => {
      // Assemble
      renderWithProviders(<Login />)
      await setupForm(['email', 'password'])

      // Act
      const loginBtn = await screen.findByRole('button', { name: 'Login' })
      await act(() => loginBtn.click())

      // Assert
      await waitFor(() => expect(mockNavigate).toBeCalledWith('/'))
    })
  })

  describe('interaction', () => {
    it('Should display error when email is invalid', async () => {
      // Assemble
      renderWithProviders(<Login />)
      await setupForm(['password'])

      // Act
      const loginBtn = await screen.findByRole('button', { name: 'Login' })
      await act(() => loginBtn.click())

      // Assert
      const emailError = await screen.findByTestId('login-form-email-error')
      expect(emailError).toBeInTheDocument()
    })

    it('Should display error when password is invalid', async () => {
      // Assemble
      renderWithProviders(<Login />)
      await setupForm(['email'])

      // Act
      const loginBtn = await screen.findByRole('button', { name: 'Login' })
      await act(() => loginBtn.click())

      // Assert
      const passwordError = await screen.findByTestId(
        'login-form-password-error',
      )
      expect(passwordError).toBeInTheDocument()
    })
  })
})
