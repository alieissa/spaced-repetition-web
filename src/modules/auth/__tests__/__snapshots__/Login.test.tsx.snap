// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login view should render correctly 1`] = `
<DocumentFragment>
  <div
    class="ui container"
  >
    <div
      class="justify-space-between bordered p-1r mb-1r"
    >
      <h1
        class="ui header"
        style="margin-top: 0px; margin-bottom: 0px;"
      >
        Spaced Reps
      </h1>
      <div>
        <button
          class="ui blue medium button"
          data-testid="login-form-signup-button"
          style="border-radius: 0;"
        >
          Signup
        </button>
      </div>
    </div>
    <div
      data-testid="login-form"
    >
      <form
        class="ui form flex-column bordered p-1r"
      >
        <div
          class="field"
        >
          <label>
            Email
          </label>
          <div
            class="ui input"
            style="opacity: 1;"
          >
            <input
              data-testid="login-form-email-input"
              placeholder="Email"
              style="border-radius: unset;"
              title="email"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="field"
        >
          <label>
            Password
          </label>
          <div
            class="ui input"
            style="opacity: 1;"
          >
            <input
              data-testid="login-form-password-input"
              placeholder="Password"
              style="border-radius: unset;"
              title="password"
              type="password"
              value=""
            />
          </div>
        </div>
        <footer
          class="align-center justify-space-between"
        >
          <div
            class="align-center"
          >
            <div
              class="ui checked fitted checkbox sp-checkbox"
            >
              <input
                checked=""
                class="hidden"
                readonly=""
                tabindex="0"
                type="checkbox"
                value=""
              />
              <label />
            </div>
            <span
              class="ml-fourth-r"
            >
              Remember Me?
            </span>
          </div>
          <button
            class="ui button align-self-end"
            style="border-radius: 0;"
            type="submit"
          >
            Login
          </button>
        </footer>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
